notify/jira/types.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2023 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jira
+
+import (
+	"encoding/json"
+	"errors"
+	"reflect"
+)
+
+type templateFunc func(string) (string, error)
+
+type issue struct {
+	Key        string       `json:"key,omitempty"`
+	Fields     *issueFields `json:"fields,omitempty"`
+	Transition *idNameValue `json:"transition,omitempty"`
+}
+
+type issueFields struct {
+	Description any           `json:"description"`
+	Issuetype   *idNameValue  `json:"issuetype,omitempty"`
+	Labels      []string      `json:"labels,omitempty"`
+	Priority    *idNameValue  `json:"priority,omitempty"`
+	Project     *issueProject `json:"project,omitempty"`
+	Components  []idNameValue `json:"components,omitempty"`
+	Resolution  *idNameValue  `json:"resolution,omitempty"`
+	Summary     string        `json:"summary"`
+	Status      *issueStatus  `json:"status,omitempty"`
+
+	CustomFields map[string]any `json:"-"`
+}
+
+type idNameValue struct {
+	ID   string `json:"id,omitempty"`
+	Name string `json:"name,omitempty"`
+}
+
+type issueProject struct {
+	Key string `json:"key"`
+}
+
+type issueStatus struct {
+	Name           string `json:"name"`
+	StatusCategory struct {
+		Key string `json:"key"`
+	} `json:"statusCategory"`
+}
+
+type issueSearch struct {
+	Expand       []string `json:"expand"`
+	Fields       []string `json:"fields"`
+	FieldsByKeys bool     `json:"fieldsByKeys"`
```

Comment:
The presence of this field fails with my JIRA instance:

```
ts=2024-01-05T15:25:57.986Z caller=dispatch.go:353 level=error component=dispatcher msg="Notify for alerts failed" num_alerts=1 err="jira/jira[0]: notify retry canceled due to unrecoverable 
error after 1 attempts: unexpected status code 400: {\"errorMessages\":[\"Unrecognized field \\\"fieldsByKeys\\\" (Class com.atlassian.jira.rest.v2.search.SearchRequestBean), not marked as i
gnorable\\n at [Source: org.apache.catalina.connector.CoyoteInputStream@148a2304; line: 1, column: 54] (through reference chain: com.atlassian.jira.rest.v2.search.SearchRequestBean[\\\"field
sByKeys\\\"])\"]}"                                                                                                                                                                            

```

It succeeds if I remove/comment this line.

Reply:
Are you testing this against an on premise instance?

Reply:
Removed `fieldsByKeys` in 3aa5a4c and tested against Cloud API